tests/unit/core/server/testServer.js

```diff
@@ -83,9 +110,151 @@ describe('core/server/base', () => {
+      const response = await testClient({ store, sagaMiddleware, appSagas: userSaga })
+        .get('/en-US/firefox/')
+        .set('cookie', `${defaultConfig.get('cookieName')}="${token}"`)
+        .end();
+
+      const { api, user } = store.getState();
+
+      expect(response.statusCode).toEqual(200);
+      expect(api.token).toEqual(token);
+      expect(user.id).toEqual(42);
+      expect(user.username).toEqual('babar');
+      mockUserApi.verify();
+    });
+
+    it('returns a 500 error page when retrieving the user profile fails', async () => {
+      mockUserApi
+        .expects('userProfile')
+        .once()
+        .returns(Promise.reject(new Error('example of an API error')));
+
+      const token = userAuthToken();
+      const { store, sagaMiddleware } = createStoreAndSagas();
+      const response = await testClient({ store, sagaMiddleware, appSagas: userSaga })
+        .get('/en-US/firefox/')
+        .set('cookie', `${defaultConfig.get('cookieName')}="${token}"`)
+        .end();
+
+      expect(response.statusCode).toEqual(500);
+    });
+
+    it('fetches the user profile even when SSR is disabled', async () => {
+      const profile = createUserProfileResponse({ id: 42, username: 'babar' });
+
+      mockUserApi
+        .expects('userProfile')
+        .once()
+        .returns(Promise.resolve(profile));
+
+      const token = userAuthToken();
+      const { store, sagaMiddleware } = createStoreAndSagas();
+      // We use `cloneDeep()` to allow modifications on the `config` object,
+      // since a call to `get()` makes it immutable. This is the case in the
+      // previous test cases (on `defaultConfig`).
+      const config = configUtil.cloneDeep(defaultConfig);
+      config.disableSSR = true;
+
+      const client = testClient({
+        store,
+        sagaMiddleware,
+        appSagas: userSaga,
+        config,
+      });
+
+      const response = await client
+        .get('/en-US/firefox/')
+        .set('cookie', `${defaultConfig.get('cookieName')}="${token}"`)
+        .end();
+
+      const { api, user } = store.getState();
+
+      expect(response.statusCode).toEqual(200);
+      expect(api.token).toEqual(token);
+      expect(user.id).toEqual(42);
+      expect(user.username).toEqual('babar');
+      mockUserApi.verify();
+    });
+
+    it('it serializes the redux state in html', async () => {
+      const profile = createUserProfileResponse({ id: 42, username: 'babar' });
+
+      mockUserApi
+        .expects('userProfile')
+        .once()
+        .returns(Promise.resolve(profile));
+
+      const token = userAuthToken();
+      const { store, sagaMiddleware } = createStoreAndSagas();
+      // We use `cloneDeep()` to allow modifications on the `config` object,
+      // since a call to `get()` makes it immutable. This is the case in the
+      // previous test cases (on `defaultConfig`).
+      const config = configUtil.cloneDeep(defaultConfig);
+      config.disableSSR = true;
```

Comment:
I don't think it's so important to test this with `disableSSR=true`. It would be better not to just because it will simulate production better.

Reply:
I see what you mean, and I will add another similar test for that.

I'd like to keep this one though as it actually covers the fact that with `disableSSR=true`, sagas are executed correctly and both user profile and token are passed to the client side. The bug you mentioned earlier this week was because of that.

Reply:
The state computed on the server is still passed to the client _via_ the HTML. `SET_AUTH_TOKEN` is not often called on the client because it is already done by the server. That is also why I put these assertions in the `disableSSR=true` test case. That is the HTML I get with `disableSSR=true`:

``` html
<!DOCTYPE html>
<html lang="en-US" dir="ltr" data-reactroot="" data-reactid="1" data-react-checksum="1242348688">
 <head data-reactid="2">
  <meta charset="utf-8" data-reactid="3" />
  <meta name="viewport" content="width=device-width, initial-scale=1" data-reactid="4" />
  <link rel="shortcut icon" href="/favicon.ico" data-reactid="5" />
  <title data-react-helmet="true" data-reactid="6"></title>
  <noscript data-reactid="7">
   <style data-reactid="8">/* !important is needed to show the button, make sure switch is hidden */
.InstallButton .InstallButton-button {
  display: inline-block !important;
}

.InstallButton .InstallButton-switch {
  display: none !important;
}
</style>
  </noscript>
 </head>
 <body data-reactid="9">
  <div id="react-view" data-reactid="10"></div>
  <script type="application/json" id="redux-store-state" data-reactid="11">{"addons":{},"api":{"clientApp":"firefox","lang":"en-US","token":"eyJhdXRoX2hhc2giOiJiNTRhNTBiYjYzMDg0OTVkMzQ3OTU5MGIyMDg4MmEyZDkwZWM3ZmJjIiwidXNlcl9pZCI6MTA2MzkxNTZ9:1dmdwc:DHZayyST3Y2Ms6SE7memWNz5l7Y","userAgent":"Mozilla\u002F5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit\u002F537.36 (KHTML, like Gecko) Chrome\u002F59.0.3071.115 Safari\u002F537.36","userAgentInfo":{"browser":{"name":"Chrome","version":"59.0.3071.115","major":"59"},"os":{"name":"Mac OS","version":"10.12.6"}}},"autocomplete":{"loading":false,"suggestions":[]},"categories":{"categories":null,"loading":false},"errors":{},"errorPage":{"clearOnNext":false,"error":null,"hasError":false,"statusCode":null},"featured":{"addonType":null,"loading":false,"results":[]},"infoDialog":{},"installations":{},"landing":{"addonType":null,"featured":{"count":0,"results":[]},"highlyRated":{"count":0,"results":[]},"loading":false,"popular":{"count":0,"results":[]},"resultsLoaded":false},"loadingBar":0,"reduxAsyncConnect":{"loaded":false,"loadState":{}},"reviews":{"byAddon":{}},"search":{"count":0,"filters":{},"loading":false,"results":[]},"user":{"id":10639156,"username":"willdurand"},"viewContext":{"context":"VIEW_CONTEXT_EXPLORE"}}</script>
  <script src="//127.0.0.1:3001/amo-c623c6e9e4951a016e4b.js" data-reactid="12"></script>
 </body>
</html>
```

I have updated both test cases now. The `disableSSR=true` one checks that the state is correct (so that login works with such a configuration). The second test case about making sure the state is serialized into HTML now simulates the production configuration.

Reply:
> I'd like to keep this one though as it actually covers the fact that with disableSSR=true, sagas are executed correctly and both user profile and token are passed to the client side.

Adding both test cases is fine but I didn't explain myself very well in my suggestion to only include a test for the `disableSSR: false` case (to emulate production). Here is some more context.

The `core/server/testServer.js` suite performs high level integration tests--that is, at the server->client level. These types of tests are good but they are also dangerous because they tempt you to test for a matrix of possibilities and that matrix can be vast. The more test cases you add, the slower the suite becomes.

I think the best approach for an app like ours is to rely on unit tests as much as possible and then try to use static analysis (like Flow) to make sure our units are glued together right. We may not ever get to that last part because I haven't been able to convince the team of how Flow can accomplish this for us.

So, we should still do integration level tests but we should use them sparingly. For this case, we actually [already have a unit test](https://github.com/mozilla/addons-frontend/blob/master/tests/unit/core/containers/TestServerHtml.js#L101) for rendering the redux state to the client. We were lacking an integration test for it so that's why I requested adding one (it could definitely catch a bug or two) but I don't think we need to try and cover every edge case.

Reply:
Oh, but you are absolutely right! Thanks for the explanations.